blas/impl/KokkosBlas1_axpby_unification_attempt_traits.hpp

```diff
@@ -0,0 +1,1009 @@
+  using type = typename T::array_layout;
+};
+
+// --------------------------------
+
+template <typename T>
+constexpr bool isTypeComplex() {
+  return (std::is_same_v<T, Kokkos::complex<float>> ||
+          std::is_same_v<T, Kokkos::complex<double>> ||
+          std::is_same_v<T, Kokkos::complex<long double>> ||
+          std::is_same_v<T, Kokkos::complex<int>> ||
+          std::is_same_v<T, Kokkos::complex<unsigned int>> ||
+          std::is_same_v<T, Kokkos::complex<long int>> ||
+          std::is_same_v<T, Kokkos::complex<unsigned long int>> ||
+          std::is_same_v<T, Kokkos::complex<size_t>> ||
+          std::is_same_v<T, Kokkos::complex<std::int32_t>> ||
+          std::is_same_v<T, Kokkos::complex<std::uint32_t>> ||
+          std::is_same_v<T, Kokkos::complex<std::int64_t>> ||
+          std::is_same_v<T, Kokkos::complex<std::uint64_t>>);
+}
+
+// --------------------------------
+
+template <class tExecSpace, class AV, class XMV, class BV, class YMV>
+struct AxpbyUnificationAttemptTraits {
+  static constexpr bool atDevCase =
+      KokkosKernels::Impl::kk_is_gpu_exec_space<tExecSpace>();
+  static constexpr bool atHostCase = !atDevCase;
+
+  static constexpr bool Asc  = !Kokkos::is_view_v<AV>;
+  static constexpr bool Ar0  = Tr0_val<AV>();
+  static constexpr bool Ar1s = Tr1s_val<AV>();
+  static constexpr bool Ar1d = Tr1d_val<AV>();
+  static constexpr bool Avi  = Ar0 || Ar1s || Ar1d;
+
+  static constexpr bool Xr1 = Kokkos::is_view_v<XMV> && (XMV::rank == 1);
+  static constexpr bool Xr2 = Kokkos::is_view_v<XMV> && (XMV::rank == 2);
+
+  static constexpr bool Bsc  = !Kokkos::is_view_v<BV>;
+  static constexpr bool Br0  = Tr0_val<BV>();
+  static constexpr bool Br1s = Tr1s_val<BV>();
+  static constexpr bool Br1d = Tr1d_val<BV>();
+  static constexpr bool Bvi  = Br0 || Br1s || Br1d;
+
+  static constexpr bool Yr1 = Kokkos::is_view_v<YMV> && (YMV::rank == 1);
+  static constexpr bool Yr2 = Kokkos::is_view_v<YMV> && (YMV::rank == 2);
+
+  static constexpr bool xyRank1Case = Xr1 && Yr1;
+  static constexpr bool xyRank2Case = Xr2 && Yr2;
+
+  // ********************************************************************
+  // In order to better understand the lines between now and right before
+  // the constructor, assume that all constructor checks.
+  // ********************************************************************
+
+  // ********************************************************************
+  // Declare 'AtInputScalarTypeA'
+  // ********************************************************************
+  using ScalarTypeA2_atDev =
+      typename getScalarTypeFromView<AV, Avi && atDevCase>::type;
+  using ScalarTypeA1_atDev =
+      std::conditional_t<Asc && atDevCase, AV, ScalarTypeA2_atDev>;
+
+  using ScalarTypeA2_atHost =
+      typename getScalarTypeFromView<AV, Avi && atHostCase>::type;
+  using ScalarTypeA1_atHost =
+      std::conditional_t<Asc && atHostCase, AV, ScalarTypeA2_atHost>;
+
+  using AtInputScalarTypeA =
```

Comment:
As far as I can tell, we could achieve the same result with:
`using AtInputScalarTypeA = std::conditional_t<Kokkos::is_view_v<AV>, typename AV::scalar_type, AV>;`
Please let me know if I'm wrong otherwise we should clean up the above to avoid confusion...